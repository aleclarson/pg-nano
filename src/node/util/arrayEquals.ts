export function arrayEquals(
  a: unknown[] | null | undefined,
  b: unknown[] | null | undefined,
): boolean {
  if (!a) {
    return !b
  }
  if (!b) {
    return false
  }
  return (
    a === b ||
    (a.length === b.length && a.every((val, index) => Object.is(val, b[index])))
  )
}
