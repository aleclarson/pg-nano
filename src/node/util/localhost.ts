/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2014 Joseph Werle
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import fs from 'node:fs'
import net from 'node:net'

export function isLocalHost(host: string): boolean {
  if (host === '::1' || host === 'localhost') {
    return true
  }

  if (process.env.DEV_CONTAINER === 'pg-nano' && host === 'postgres') {
    return true
  }

  if (net.isIP(host)) {
    for (const line of readHostsFile()) {
      if (/^(127\.0\.0\.1|::1)/.test(line.trim())) {
        const hosts = line.split(/\s+/)
        if (hosts.includes(host)) {
          return true
        }
      }
    }
    return false
  }

  // 127.0.0.1 - 127.255.255.254
  const parts = host.split('.').map(Number)
  if (parts.length !== 4) {
    return false
  }
  if (parts[0] !== 127) {
    return false
  }
  if (parts[1] > 255 || parts[1] < 0) {
    return false
  }
  if (parts[2] > 255 || parts[2] < 0) {
    return false
  }
  if (parts[3] > 254 || parts[3] < 1) {
    return false
  }
  return true
}

function readHostsFile(): string[] {
  let configs: string
  try {
    configs = fs.readFileSync('/etc/hosts', 'utf-8')
  } catch {
    return []
  }
  return configs.split('\n')
}
