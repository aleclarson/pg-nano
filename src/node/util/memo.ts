export interface MemoOptions<Fn extends (...args: any[]) => any> {
  cache?: Record<string, ReturnType<Fn>>
  key?: (...args: Parameters<Fn>) => string
}

/**
 * Creates a memoized function. The returned function will only
 * execute the source function when no value has previously been
 * computed.
 *
 * @see https://radashi-org.github.io/reference/curry/memo
 * @example
 * ```ts
 * const calls: number[] = []
 * const fib = memo((x: number) => {
 *   calls.push(x)
 *   return x < 2 ? x : fib(x - 1) + fib(x - 2)
 * })
 *
 * fib(10) // 55
 * fib(10) // 55
 * // calls === [10]
 *
 * fib(11) // 89
 * // calls === [10, 11]
 * ```
 */
export function memo<Fn extends (...args: any[]) => any>(
  func: Fn,
  options: MemoOptions<Fn> = {},
): Memoized<Fn> {
  const { key: keyFunc, cache = {} } = options

  return function callWithMemo(...args) {
    const key = keyFunc ? keyFunc(...args) : JSON.stringify({ args })
    if (key in cache) {
      return cache[key]
    }
    const result = func(...args)
    cache[key] = result
    return result
  } as Memoized<Fn>
}

export type Memoized<Fn extends (...args: any[]) => any> =
  ReturnType<Fn> extends infer Result
    ? (...args: Parameters<Fn>) => Result
    : never
