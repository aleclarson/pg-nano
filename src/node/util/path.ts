import path from 'node:path'

export function cwdRelative(p: string) {
  const rel = path.relative(process.cwd(), p)
  return rel.startsWith('..') ? rel : './' + rel
}
