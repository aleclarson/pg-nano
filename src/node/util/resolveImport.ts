export function resolveImport(specifier: string) {
  // Once Vite supports import.meta.resolve, we can remove the require.resolve
  // fallback. See: https://github.com/vitejs/vite/discussions/15871
  let resolvedPath: string
  if (typeof import.meta.resolve === 'function') {
    resolvedPath = new URL(import.meta.resolve(specifier)).pathname
  } else {
    resolvedPath = require.resolve(specifier)
  }

  return resolvedPath
}
