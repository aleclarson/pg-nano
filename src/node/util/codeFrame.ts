import { cwdRelative } from './path.js'

export function appendCodeFrame(
  error: Error,
  errorPosition: number,
  query: string,
  stmtLine = 1,
  stmtFile?: string,
) {
  const errorLine = query.slice(0, errorPosition).split('\n').length - 1
  const errorColumn =
    errorPosition - (1 + query.lastIndexOf('\n', errorPosition - 1))

  const queryLines = (query + ';').split('\n')
  const startLine = Math.max(0, errorLine - 2)
  const endLine = Math.min(queryLines.length - 1, errorLine + 3)
  const width = String(stmtLine + endLine).length

  let output = ''

  for (let line = startLine; line <= endLine; line++) {
    output += `${line === errorLine ? '> ' : '  '}${String(stmtLine + line).padStart(width)} | ${queryLines[line]}\n`

    if (line === errorLine) {
      output += ' '.repeat(3 + width) + `| ${' '.repeat(errorColumn - 1)}^\n`
    }
  }

  error.message =
    error.message.replace(/\nLINE \d+: [\s\S]+$/, '') +
    '\n\n' +
    output +
    (stmtFile
      ? '\n    at ' +
        cwdRelative(stmtFile) +
        ':' +
        (stmtLine + errorLine) +
        ':' +
        errorColumn
      : '')
}
