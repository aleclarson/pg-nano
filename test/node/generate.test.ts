import { dedent, TestProject } from '../util.js'

const sql = dedent

describe('generate', () => {
  test('zero argument routine', async () => {
    const project = new TestProject({
      fixtures: {
        'sql/schema.sql': sql`
          CREATE FUNCTION forty_two() RETURNS int AS $$
          BEGIN
            RETURN 42;
          END;
          $$ LANGUAGE plpgsql;
        `,
      },
    })

    await project.update()
    expect(project.readGeneratedFiles()).toMatchSnapshot()

    type Schema =
      typeof import('../__fixtures__/generate__zero_argument_routine/sql/schema.ts')

    const client = await project.importClient<Schema>()

    expect(await client.fortyTwo()).toBe(42)
  })

  test('one argument routine', async () => {
    const project = new TestProject({
      fixtures: {
        'sql/schema.sql': sql`
          CREATE FUNCTION multiply_by_2(x int) RETURNS int AS $$
          BEGIN
            RETURN x * 2;
          END;
          $$ LANGUAGE plpgsql;
        `,
      },
    })

    await project.update()
    expect(project.readGeneratedFiles()).toMatchSnapshot()

    type Schema =
      typeof import('../__fixtures__/generate__one_argument_routine/sql/schema.ts')

    const client = await project.importClient<Schema>()

    expect(await client.multiplyBy2(21)).toBe(42)
  })
})
