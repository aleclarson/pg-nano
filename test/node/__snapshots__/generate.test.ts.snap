// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generate > one argument routine 1`] = `
{
  "sql/schema.ts": "/* BEWARE: This file was generated by pg-nano. Any changes you make will be overwritten. */
import { bindQueryValue } from 'pg-nano'

export declare namespace multiplyBy2 {
  type Params = { x: number }
  type Result = number
}

export const multiplyBy2 = /* @__PURE__ */ bindQueryValue<multiplyBy2.Params, multiplyBy2.Result>("multiply_by_2", $ => $.arity(1, 1).namedArgs(["x"]))
",
  "sql/typeData.ts": null,
}
`;

exports[`generate > zero argument routine 1`] = `
{
  "sql/schema.ts": "/* BEWARE: This file was generated by pg-nano. Any changes you make will be overwritten. */
import { bindQueryValue } from 'pg-nano'

export declare namespace fortyTwo {
  type Params = []
  type Result = number
}

export const fortyTwo = /* @__PURE__ */ bindQueryValue<fortyTwo.Params, fortyTwo.Result>("forty_two", $ => $.arity(0, 0))
",
  "sql/typeData.ts": null,
}
`;
